core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -124,6 +124,25 @@ class EntityService(protected val userInfo: UserInfo, val dataSource: SlickDataS
       }
     }
 
+  def renameEntityType(workspaceName: WorkspaceName, oldName: String, newName: String): Future[Int] =
+    getWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.write, Some(WorkspaceAttributeSpecs(all = false))) flatMap { workspaceContext =>
+      dataSource.inTransaction { dataAccess =>
+        dataAccess.entityQuery.doesEntityTypeAlreadyExist(workspaceContext, newName) flatMap {
+          _.head match {
+            case false =>
+              dataAccess.entityQuery.doesEntityTypeAlreadyExist(workspaceContext, oldName) flatMap {
+                _.head match {
+                  case false => throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.NotFound,
+                    s"Can't find entity type ${oldName}"))
+                  case true => dataAccess.entityQuery.changeEntityTypeName(workspaceContext, oldName, newName)
+                }
+              }
+            case true => throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.Conflict, s"${newName} already exists as an entity type"))
+          }
+        }
+      }
+    }
```

Comment:
you're looking for tips on how to clean up this code … my suggestion is to modularize the logic around validating the existence or non-existence of the oldName and the newName, by creating separate functions to handle those. With that in place, you could move to a for-yield instead of the flatMaps.

Here's an example of such a refactor. NOTE: this refactor assumes that `doesEntityTypeAlreadyExist` is rewritten using `uniqueResult` to return an `Option` instead of a `Seq` (see my previous comment):

```scala
  def renameEntityType(workspaceName: WorkspaceName, oldName: String, newName: String): Future[Int] = {

    import org.broadinstitute.dsde.rawls.dataaccess.slick.{DataAccess, ReadAction}

    def validateExistingType(dataAccess: DataAccess, workspaceContext: Workspace, oldName: String): ReadAction[Boolean] = {
      dataAccess.entityQuery.doesEntityTypeAlreadyExist(workspaceContext, oldName) map {
        case Some(true) => true
        case Some(false) => throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.NotFound,
          s"Can't find entity type ${oldName}"))
        case None =>  throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.InternalServerError,
          s"Unexpected error; could not determine existence of entity type ${oldName}"))
      }
    }

    def validateNewType(dataAccess: DataAccess, workspaceContext: Workspace, newName: String): ReadAction[Boolean] = {
      dataAccess.entityQuery.doesEntityTypeAlreadyExist(workspaceContext, newName) map {
        case Some(true) => throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.Conflict, s"${newName} already exists as an entity type"))
        case Some(false) => false
        case None =>  throw new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.InternalServerError,
          s"Unexpected error; could not determine existence of entity type ${newName}"))
      }
    }

    getWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.write, Some(WorkspaceAttributeSpecs(all = false))) flatMap { workspaceContext =>
      dataSource.inTransaction { dataAccess =>
        for {
          _ <- validateNewType(dataAccess, workspaceContext, newName)
          _ <- validateExistingType(dataAccess, workspaceContext, oldName)
          renameResult <- dataAccess.entityQuery.changeEntityTypeName(workspaceContext, oldName, newName)
        } yield {
          renameResult
        }
      }
    }
  }
```

I haven't run this code through unit tests, but hopefully it illustrates the point

Reply:
I've incorporated this suggestion in its totality.